avocado/core/status/repo.py

```diff
@@ -0,0 +1,63 @@
+from .utils import json_loads
+
+
+class StatusMsgMissingDataError(Exception):
+    """Status message does not contain the required data."""
+
+
+class StatusRepo:
+    """Maintains tasks' status related data and provides aggregated info."""
+
+    def __init__(self):
+        self._data = {}
+        self._by_result = {}
+
+    def _set_by_result(self, message):
+        """Sets an entry in the aggregate by result.
+
+        For messages that include a "result" key, expected for example,
+        from a "finished" status message, this will allow users to query
+        for tasks with a given result."""
+        result = message.get('result')
+        if result not in self._by_result:
+            self._by_result[result] = []
+        self._by_result[result].append(message['id'])
+
+    def _handle_task_running(self, message):
+        self._set_task_data(message)
+
+    def _handle_task_started(self, message):
+        if 'output_dir' not in message:
+            raise StatusMsgMissingDataError('output_dir')
+        self._set_task_data(message)
+
+    def _handle_task_finished(self, message):
+        self._set_by_result(message)
+        self._set_task_data(message)
+
+    def process_raw_message(self, raw_message):
+        raw_message = raw_message.strip()
+        message = json_loads(raw_message)
+        self.process_message(message)
+
+    def process_message(self, message):
+        if 'id' not in message:
+            raise StatusMsgMissingDataError('id')
+
+        if message.get('status') == 'running':
```

Comment:
I'm sending a slight variation of this proposal because I've realized that:

1) `running` currently doesn't needs any special processing
2) the default action should be just to set the data (not raise an error)

Something like:

```python
    def process_message(self, message):
        if 'id' not in message:
            raise StatusMsgMissingDataError('id')

        handlers = {'started': self._handle_task_started,
                    'finished': self._handle_task_finished}
        meth = handlers.get(message.get('status'),
                            self._set_task_data)
        meth(message)

```